### PR TITLE
Anticipate changes introduced by new proxy version

### PR DESCRIPTION
**Pull request by sboeuf**
The purpose of this PR is to bring the changes we will need to integrate very quickly the new proxy version. Once the proxy will be released, only very minor changes will be needed to get virtcontainers working with this new version.

### CODE REVIEW

File: cc_proxy.go [hunk 1]
```diff
@@ -49,65 +49,63 @@ func (p *ccProxy) connectProxy(runtimeSocketPath string) (*api.Client, error) {
...
+		return []ProxyInfo{}, fmt.Errorf("Wrong agent config type, should be HyperConfig type")
 	}
 
 	_, err = p.client.Hello(pod.id, hyperConfig.SockCtlName, hyperConfig.SockTtyName, nil)
 	if err != nil {
-		return []IOStream{}, err
+		return []ProxyInfo{}, err
+	}
+
+	if newInfos == false {
```
Comment (jodh-intel) [inline cc_proxy.go]: This name is in keeping with `ProxyInfo`, but I found it rather unclear. If I'm understanding it, a comment to state that this parameter controls whether a new set of proxy I/O streams is allocated, or whether the existing ones are used would have made it much clearer I think (or maybe renaming it `allocateNewStreams` or similar).

Comment (sboeuf) [inline cc_proxy.go]: Well we are not allocating new streams with the new proxy, we are only getting tokens, and I didn't want to use a too specific name like "token". That's why I used info.

Comment (sboeuf) [inline cc_proxy.go]: maybe newProxyInfo ?

Comment (jodh-intel) [inline cc_proxy.go]: Yeah, or maybe `newConnectionInfo` (well, that's a bit long, but... ;-)

Maybe you could invert the logic and have:

```
reuseConnection bool
```

... as that's clearer still I think.

Comment (sboeuf) [inline cc_proxy.go]: yes reuseConnection sounds good :)

**Comment by jodh-intel** [comment]
lgtm

File: container.go [hunk 2]
```diff
@@ -233,20 +233,27 @@ func createContainer(pod *Pod, contConfig ContainerConfig) (*Container, error) {
 		return nil, err
 	}
 
+	process, err := c.pod.storage.fetchContainerProcess(c.podID, c.id)
```
Comment (sameo) [inline container.go]: `c.fetchProcess()`

File: cc_proxy.go [hunk 3]
```diff
@@ -49,65 +49,63 @@ func (p *ccProxy) connectProxy(runtimeSocketPath string) (*api.Client, error) {
...
+		return []ProxyInfo{}, fmt.Errorf("Wrong agent config type, should be HyperConfig type")
 	}
 
 	_, err = p.client.Hello(pod.id, hyperConfig.SockCtlName, hyperConfig.SockTtyName, nil)
 	if err != nil {
-		return []IOStream{}, err
+		return []ProxyInfo{}, err
+	}
+
+	if reuseConnection == true {
```
Comment (sboeuf) [inline cc_proxy.go]: Agreed

File: agent.go [hunk 4]
```diff
@@ -118,29 +118,29 @@ type agent interface {
 	init(pod *Pod, config interface{}) error
 
 	// startAgent will start the agent.
-	startAgent() error
+	startAgent(pod *Pod) error
```
Comment (sboeuf) [inline agent.go]: Agreed, I'll do the same for stopAgent() ==> stop()

File: cc_proxy.go [hunk 5]
```diff
@@ -49,65 +49,67 @@ func (p *ccProxy) connectProxy(runtimeSocketPath string) (*api.Client, error) {
...
 	_, err = p.client.Hello(pod.id, hyperConfig.SockCtlName, hyperConfig.SockTtyName, nil)
 	if err != nil {
-		return []IOStream{}, err
+		return []ProxyInfo{}, err
+	}
+
+	// reuseConnection is intended to be true in case we don't want
+	// the proxy to create new tokens for every container described
+	// in the pod, but instead only connect to the agent inside the
+	// VM.
```
Comment (sameo) [inline cc_proxy.go]: That should only happen with `connect`, I think we can remove that one.

File: proxy.go [hunk 6]
```diff
@@ -89,30 +88,38 @@ func newProxyConfig(config PodConfig) interface{} {
...
 	// register connects and registers the proxy to the given VM.
-	// It also returns streams related to containers workloads.
-	register(pod Pod) ([]IOStream, error)
+	// It also returns information related to containers workloads.
+	//
+	// reuseConnection is intended to be true in case we don't want
+	// the proxy to create new tokens for every container described
+	// in the pod, but instead only connect to the agent inside the
+	// VM.
+	register(pod Pod, reuseConnection bool) ([]ProxyInfo, error)
```
Comment (sameo) [inline proxy.go]: You can remove that one.

File: proxy.go [hunk 7]
```diff
@@ -89,30 +88,38 @@ func newProxyConfig(config PodConfig) interface{} {
...
 
 	// connect gets the proxy a handle to a previously registered VM.
-	// It also returns streams related to containers workloads.
-	connect(pod Pod) (IOStream, error)
+	// It also returns information related to containers workloads.
+	//
+	// reuseConnection is intended to be true in case we don't want
+	// the proxy to create a new token, but instead only get a handle
+	// to be able to communicate with the agent inside the VM.
+	connect(pod Pod, reuseConnection bool) (ProxyInfo, error)
```
Comment (sameo) [inline proxy.go]: I personally find that passing a `createToken bool` would be more logical.

**Comment by sameo** [comment]
LGTM

**Pull request merged**